core/src/utils/sync/racy_lock.rs

```diff
@@ -0,0 +1,174 @@
+/// with the same interface as `std::sync::LazyLock`.
+///
+/// The underlying implementation is based on `once_cell::sync::race::OnceBox` which relies on
+/// `core::atomic::AtomicPtr` to ensure that the data race results in a single successful
+/// write to the relevant pointer, namely the first write.
+/// See <https://github.com/matklad/once_cell/blob/v1.19.0/src/race.rs#L294>.
+///
+/// Performs lazy evaluation and can be used for statics.
+pub struct RacyLock<T, F = fn() -> T>
+where
+    F: Fn() -> T,
+{
+    inner: AtomicPtr<T>,
+    f: F,
+}
+
+impl<T, F> RacyLock<T, F>
+where
+    F: Fn() -> T,
+{
+    /// Creates a new lazy, racy value with the given initializing function.
+    pub const fn new(f: F) -> Self {
+        Self {
+            inner: AtomicPtr::new(ptr::null_mut()),
+            f,
+        }
+    }
+
+    /// Forces the evaluation of the locked value and returns a reference to
+    /// the result. This is equivalent to the `Deref` impl, but is explicit.
+    ///
+    /// This method will not block the calling thread if another initialization
+    /// routine is currently running.
+    pub fn force(this: &RacyLock<T, F>) -> &T {
+        let mut ptr = this.inner.load(Ordering::Acquire);
+
+        // Pointer is not yet set, attempt to set it ourselves.
+        if ptr.is_null() {
+            // Execute the initialization function and allocate.
+            let val = (this.f)();
+            ptr = Box::into_raw(Box::new(val));
+
+            // Attempt atomic store.
+            let exchange = this.inner.compare_exchange(
+                ptr::null_mut(),
+                ptr,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            );
+
+            // Pointer already set, load.
+            if let Err(old) = exchange {
+                drop(unsafe { Box::from_raw(ptr) });
+                ptr = old;
+            }
+        }
+
+        unsafe { &*ptr }
+    }
+}
+
+impl<T: Default> Default for RacyLock<T> {
+    /// Creates a new lock that will evaluate the underlying value based on `T::default`.
+    #[inline]
+    fn default() -> RacyLock<T> {
+        RacyLock::new(T::default)
+    }
+}
+
+impl<T, F> fmt::Debug for RacyLock<T, F>
+where
+    T: fmt::Debug,
+    F: Fn() -> T,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "RacyLock({:?})", self.inner.load(Ordering::Relaxed))
+    }
+}
+
+impl<T, F> Deref for RacyLock<T, F>
+where
+    F: Fn() -> T,
+{
+    type Target = T;
+
+    /// Dereferences the value.
+    ///
+    /// This method will not block the calling thread if another initialization
+    /// routine is currently running.
+    #[inline]
+    fn deref(&self) -> &T {
+        RacyLock::force(self)
+    }
+}
+
+impl<T, F> Drop for RacyLock<T, F>
+where
+    F: Fn() -> T,
+{
+    fn drop(&mut self) {
+        let ptr = *self.inner.get_mut();
+        if !ptr.is_null() {
+            drop(unsafe { Box::from_raw(ptr) });
+        }
```

Comment:
I see that this was copy/pasted from `OnceBox::drop()`, but isn't there a possible double-free here? Basically, 2 threads could enter the `!ptr.is_null()` if, and call drop on the `Box`.

I'm not sure what would be the safe way to drop yet, but at least if we look at [how it's done with `Arc`](https://doc.rust-lang.org/1.80.1/src/alloc/sync.rs.html#2459-2509), it's a lot more complex (and clearly takes thread synchronization into consideration).

Reply:
Ah, but this lock is not `Clone`! So thinking of it like an `Arc` is wrong. IIUC, the only case then that 2 threads case race is if `RacyLock` is used in a static variable, but in the case of a static variable, `Drop` is never called. Hence, `Drop` is only ever called when a single thread has access to the lock, and synchronization in `drop()` is not needed.

If this is correct, can we add this explanation to a `Safety` comment above the `unsafe` block?

Reply:
`Drop` requires a mutable reference, so no other references can exist, and thus there can be no race.

Reply:
This is true of the `RacyLock` value, but my concern was about a data race on the pointer wrapped in `self.inner` (i.e. what's manipulated in the `unsafe` block, and then explicitly dropped).

I still think it's worth adding a `Safety` comment, as the reason why we don't need any synchronization in `drop()` is nontrivial.

Reply:
Agreed!

Reply:
As discussed offline, @sergerad could you add this as a `# Safety` section in the documentation for the `drop()` method (as opposed to directly above the `unsafe{}` block? The reasoning being that e.g. if we were to implement `Clone` on `RacyLock`, it would break the entire `drop()` implementation (and hence a `Safety` comment at the method level makes sense even if the method itself isn't marked unsafe).

Reply:
Agreed that the `# Safety` block in the method docstring could mislead users of the method.

My only goal here is to point future *reviewers of the implementation* in the right direction as to why the `drop()` implementation will never result in a double-free. The assumption is that most reviewers (myself included) will probably not use `AtomicPtr` often. For example, if you follow my stream of consciousness at the beginning of this thread, I mistakenly assumed `AtomicPtr` worked like an `Arc` (i.e. new owners of the underlying pointer can be created by cloning the `AtomicPtr`), and so in that model, the `drop()` implementation could lead to a double-free. I only later realized that my model of what an `AtomicPtr` is was incorrect - the only scenario where we can have multiple owners of the underlying pointer is when the `AtomicPtr` is used in a static variable. Otherwise, e.g. when used as a stack variable, it can be moved to other threads, but never cloned, so you'll only ever have one owner (at least that's my current understanding). But [static variables never have `drop()` called](https://doc.rust-lang.org/reference/items/static-items.html), and so in the only scenario where there could be multiple owners, `drop()` is never called - hence, you can assume a single owner in `drop()` and forego any synchronization.

To reformulate, I want to make it easier for future reviewers (myself included) to get to that reasoning, assuming they don't regularly think about how `AtomicPtr`s work, whether or not they are cloneable, and how that affects the correctness of the `drop()` implementation.

So taking everything into consideration, I would go back to my previous suggestion of adding a safety block above [this line](https://github.com/sergerad/miden-vm/blob/0f07bd25b36c0edccb5a1d4923f47f548355dd17/core/src/utils/sync/racy_lock.rs#L112) and give a more concise version of the above explanation as to why that unsafe block is used safely. For example,

```rs
// SAFETY: for any given value of `ptr`, we are guaranteed to have at most a single instance of
// `RacyLock` holding that value. Hence, synchronizing threads in `drop()` is not necessary, and we
// are guaranteed never to double-free. In short, since `RacyLock` doesn't implement `Clone`, the
// only scenario where there can be multiple instances of `RacyLock` across multiple threads
// referring to the same `ptr` value is when `RacyLock` is used in a static variable - but `drop()`
// is never called for static variables.
drop(unsafe { Box::from_raw(ptr) });
```

I personally prefer to be a bit too wordy rather than not, since a bit more context is likely to help the reviewer that's not very familiar with `AtomicPtr` and lock implementations, while the seasoned reviewer can just skip the comment.

Reply:
Updated, TY. I omitted the last sentence about the static variables because (purely hypothetically) the invocation of a static variable's `drop()` would not be an issue because there is only ever a single such variable.

Reply:
Yup, agreed, your comment is indeed more accurate.

However I noticed that you put it in the method's docstring instead of over the `unsafe{}` block, and in your previous comment you mentioned that you were concerned that a `# Safety` block in the docstring might be confusing to users (which I agree with you). Feel free to move it inside the method if you prefer that - I'll approve the PR regardless. Up to you!

Reply:
Fixed, thanks!